### PR TITLE
Blacklist Peer if they fail Handshake too many times

### PR DESCRIPTION
**Pull request by nisdas**
- [x] Disconnects with any peer which has failed the handshake more than 3 times.
- [x] Drops all gossipsub messages received from the peer.

### CODE REVIEW

File: beacon-chain/sync/rpc_status_test.go [hunk 1]
```diff
@@ -287,3 +287,80 @@ func TestStatusRPCRequest_RequestSent(t *testing.T) {
...
+		}
+	})
+
+	p1.AddConnectionHandler(r.sendRPCStatusRequest)
+	p1.Connect(p2)
+
+	if testutil.WaitTimeout(&wg, time.Second) {
+		t.Fatal("Did not receive stream within 1 sec")
+	}
+	time.Sleep(time.Second / 10)
```
Comment (prestonvanloon) [inline beacon-chain/sync/rpc_status_test.go]: "Don't make me think"

```suggestion
	time.Sleep(100 * time.Millisecond)
```

File: beacon-chain/sync/peerstatus/peer_statuses.go [hunk 2]
```diff
@@ -60,10 +73,44 @@ func Keys() []peer.ID {
...
+	if !ok {
+		return
+	}
+	count++
+	failureCount[pid] = count
+}
+
+// FailureCount returns the failure count for the particular peer.
+func FailureCount(pid peer.ID) int {
+	count, ok := failureCount[pid]
```
Comment (prestonvanloon) [inline beacon-chain/sync/peerstatus/peer_statuses.go]: Needs read lock

File: beacon-chain/sync/peerstatus/peer_statuses.go [hunk 3]
```diff
@@ -60,10 +73,44 @@ func Keys() []peer.ID {
 func LastUpdated(pid peer.ID) time.Time {
 	lock.RLock()
 	defer lock.RUnlock()
-	return lastUpdated[pid]
+	return peerStatuses[pid].lastUpdated
+}
+
+// BumpFailureCount increases the failure count for the particular peer.
+func BumpFailureCount(pid peer.ID) {
```
Comment (prestonvanloon) [inline beacon-chain/sync/peerstatus/peer_statuses.go]: Can we use the verb "increase"?

File: beacon-chain/p2p/handshake.go [hunk 4]
```diff
@@ -16,6 +17,16 @@ func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer
 		ConnectedF: func(net network.Network, conn network.Conn) {
 			// Must be handled in a goroutine as this callback cannot be blocking.
 			go func() {
+				if peerstatus.IsBadPeer(conn.RemotePeer()) {
+					// Add Peer to gossipsub blacklist
+					s.pubsub.BlacklistPeer(conn.RemotePeer())
+					log.Debug("Disconnecting with bad peer")
```
Comment (prestonvanloon) [inline beacon-chain/p2p/handshake.go]: please add `WithField("peerID", conn.RemotePeer().Pretty())`

File: beacon-chain/sync/peerstatus/peer_statuses.go [hunk 5]
```diff
@@ -12,30 +12,43 @@ import (
...
+	status      *pb.Status
+	lastUpdated time.Time
+}
 
 // Get most recent status from peer in cache. Threadsafe.
 func Get(pid peer.ID) *pb.Status {
 	lock.RLock()
 	defer lock.RUnlock()
-	return peerStatuses[pid]
+	return peerStatuses[pid].status
```
Comment (prestonvanloon) [inline beacon-chain/sync/peerstatus/peer_statuses.go]: This could cause nil pointer errors

File: beacon-chain/sync/peerstatus/peer_statuses.go [hunk 6]
```diff
@@ -60,10 +73,44 @@ func Keys() []peer.ID {
 func LastUpdated(pid peer.ID) time.Time {
 	lock.RLock()
 	defer lock.RUnlock()
-	return lastUpdated[pid]
+	return peerStatuses[pid].lastUpdated
```
Comment (prestonvanloon) [inline beacon-chain/sync/peerstatus/peer_statuses.go]: Please check this exists before accessing a struct member

File: beacon-chain/sync/peerstatus/peer_statuses.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/sync/peerstatus/peer_statuses.go]: the usage of this method is right after 
```
	for _, k := range peerstatus.Keys() {
```
if not we would have to return an error for this method. The same applies to the `LastUpdated` method.

**Pull request merged**